### PR TITLE
HTML API: Handle area, br, embed, keygen, wbr

### PR DESCRIPTION
**Pull request by sirreal**
Add "area", "br", "embed", "img", "keygen", "wbr" tag handling

> A start tag whose tag name is one of: "area", "br", "embed", "img", "keygen", "wbr"
> [Reconstruct the active formatting elements](https://html.spec.whatwg.org/multipage/parsing.html#reconstruct-the-active-formatting-elements), if any.
> [Insert an HTML element](https://html.spec.whatwg.org/multipage/parsing.html#insert-an-html-element) for the token. Immediately pop the [current node](https://html.spec.whatwg.org/multipage/parsing.html#current-node) off the [stack of open elements](https://html.spec.whatwg.org/multipage/parsing.html#stack-of-open-elements).
> [Acknowledge the token's self-closing flag](https://html.spec.whatwg.org/multipage/parsing.html#acknowledge-self-closing-flag), if it is set.
> Set the [frameset-ok flag](https://html.spec.whatwg.org/multipage/parsing.html#frameset-ok-flag) to "not ok".

This exception is currently unsupported, `<br>` is handled, but the special exception for `</br>` is not:

> An end tag whose tag name is "br"
> [Parse error](https://html.spec.whatwg.org/multipage/parsing.html#parse-errors). Drop the attributes from the token, and act as described in the next entry; i.e. act as if this was a "br" start tag token with no attributes, rather than the end tag token that it actually is.

Trac ticket: https://core.trac.wordpress.org/ticket/60283

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 1]
```diff
@@ -1685,7 +1698,13 @@ public static function is_void( $tag_name ) {
...
 			'SOURCE' === $tag_name ||
 			'TRACK' === $tag_name ||
-			'WBR' === $tag_name
+			'WBR' === $tag_name ||
+
+			// Obsolete
+			//
+			// This does not appear in https://html.spec.whatwg.org/#void-elements
+			// but it was a void tag and browsers treat it as such.
+			'KEYGEN' === $tag_name
```
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php]: [Demo](https://software.hixie.ch/utilities/js/live-dom-viewer/?%3Ckeygen%3Esee%2C%20it%27s%20void%3Cdiv%3Ethis%20is%20not)

Screenshot 2024-01-18 at 16 37 32

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 2]
```diff
@@ -934,12 +934,28 @@ private function step_in_body() {
...
 
+			/*
+			 * > An end tag whose tag name is "br"
+			 * >   Parse error. Drop the attributes from the token, and act as described in the next
+			 * >   entry; i.e. act as if this was a "br" start tag token with no attributes, rather
+			 * >   than the end tag token that it actually is.
+			 */
+			case '-BR':
+				$this->last_error = self::ERROR_UNSUPPORTED;
+				throw new WP_HTML_Unsupported_Exception( 'Closing BR tags require unimplemented special handling.' );
```
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php:945]: We can deal with this later, it's tricky 🙂

**Comment by sirreal** [comment]
@dmsnell Will you review please?

**Comment by sirreal** [comment]
I compared this with the html5lib-tests branch (#5794) and this includes 4 additional tests, no failures.

**Comment by sirreal** [comment]
KEYGEN has special handling in "in select" mode:

https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inselect

File: tests/phpunit/tests/html-api/wpHtmlProcessorSemanticRules.php [hunk 3]
```diff
@@ -392,4 +392,29 @@ public function test_in_body_any_other_end_tag_with_unclosed_non_special_element
...
+	 * This tag closer has special rules and support shouldn't be added without implementing full support.
+	 *
+	 * > An end tag whose tag name is "br"
+	 * >   Parse error. Drop the attributes from the token, and act as described in the next entry;
+	 * >   i.e. act as if this was a "br" start tag token with no attributes, rather than the end
+	 * >   tag token that it actually is.
+	 *
+	 * When this handling is implemented, this test should be removed and `test_br_end_tag_special_behavior`
+	 * should not be marked incomplete. It's not incorporated into the existing unsupported tag behavior test
+	 * because the opening tag is supported; only the closing tag isn't.
```
Comment (sirreal) [inline tests/phpunit/tests/html-api/wpHtmlProcessorSemanticRules.php]: This mentions an incomplete test (by name `test_br_end_tag_special_behavior`) that we've removed. Let's remove the specific reference to that test.

**Review by sirreal — COMMENTED** [review]
Thanks for the fixes @dmsnell! Apart from one comment they look good.

File: tests/phpunit/tests/html-api/wpHtmlProcessorSemanticRules.php (hunk 3 above)
Comment (dmsnell) [reply tests/phpunit/tests/html-api/wpHtmlProcessorSemanticRules.php]: thanks! good spotting. I've removed it.

**Comment by dmsnell** [comment]
Merged in https://core.trac.wordpress.org/changeset/57316
91e51f92a8c2c5e81b2a08cf2600615612cc5ab5

**Pull request closed without merging**